pytorch_transformers/modeling_openai.py

```diff
@@ -460,6 +456,13 @@ def __init__(self, config):
         self.drop = nn.Dropout(config.embd_pdrop)
         self.h = nn.ModuleList([Block(config.n_ctx, config, scale=True) for _ in range(config.n_layer)])
 
+        if hasattr(config, "pruned_heads"):
```

Comment:
I think we can move this duplicated method at the base class level and call it at the end of `PretrainedModel.from_pretrained()`, using the fact that people will re-loaded a saved model using `from_pretrained()` anyway.
What do you think?

Reply:
That sounds great but we'd need a way to identify the different ways the self-attention layers are referenced by. Right now it is different across BERT/XLM/OpenAI models:

Accessing the self-attention layer size with GPT/GPT-2: `self.h[int(layer)].attn.n_head`
With XLM: `self.attentions[int(layer)].n_heads `
With BERT: `self.encoder.layer[int(layer)].attention.self.num_attention_heads`

Reply:
We can just call the `prune_head` method of the model which is here to take care of this.
Here is how we could make this less redundant:
- add a base method `init_weights` which will take care of pruning the head when loading a configuration with pruned head. This method calls at the end a `_init_weights` method specific to each model for weights initialization. The general method could actually also handle general initialization of Linear and LayerNorm layer which are redundant across models. 
- in each attention head we just store pruned head to be sure we don't prune them two time like you did.

You need help I can take care of updating to this scenario.